internal/rest/leveldbreceipts.go

```diff
@@ -0,0 +1,373 @@
+			key := itr.Key()
+			if !strings.HasPrefix(key, "z") {
+				// we have iterated all the composite key entries
+				break
+			}
+			val := itr.Value()
+			receipt := make(map[string]interface{})
+			err := json.Unmarshal(val, &receipt)
+			if err != nil {
+				log.Errorf("Failed to decode stored receipt for request ID %s\n", itr.Key())
+				continue
+			} else {
+				receipt["_sequenceKey"] = key
+				results = append(results, receipt)
+			}
+		}
+		index++
+	}
+	return &results, nil
+}
+
+// getReply handles a HTTP request for an individual reply
+func (l *levelDBReceipts) GetReceipt(requestID string) (*map[string]interface{}, error) {
+	val, err := l.store.Get(requestID)
+	if err != nil {
+		if err.Error() == "leveldb: not found" {
+			return nil, nil
+		} else {
+			log.Errorf("Failed to retrieve the entry for the original key: %s. %s\n", requestID, err)
+			return nil, fmt.Errorf("Failed to the entry for the original key: %s. %s", requestID, err)
+		}
+	}
+	// returned val represents the composite key, use it to retrieve the actual content
+	lookupKey := string(val)
+	content, err := l.store.Get(lookupKey)
+	if err != nil {
+		log.Errorf("Failed to retrieve the entry using the generated ID: %s. %s\n", lookupKey, err)
+		return nil, fmt.Errorf("Failed to retrieve the entry using the generated ID: %s. %s", requestID, err)
+	}
+
+	result := make(map[string]interface{})
+	err = json.Unmarshal(content, &result)
+	if err != nil {
+		return nil, err
+	}
+	return &result, nil
+}
+
+func (l *levelDBReceipts) findEndPoint(sinceEpochMS int64) *util.Range {
+	searchKey := fmt.Sprintf("receivedAt:%d:", sinceEpochMS)
+	itr := l.store.NewIterator()
+	defer itr.Release()
+	found := itr.Seek(searchKey)
+	if found {
+		foundKey := itr.Key()
+		log.Infof("Located first-hit entry for search key %s: %s", searchKey, foundKey)
+		// must check if the found entry is a "receivedAt" index, because we may have gone past
+		// all the "receiveAt" entries without a hit
+		if strings.HasPrefix(foundKey, "receivedAt:") {
+			segments := strings.Split(foundKey, ":")
+			return &util.Range{
+				Limit: []byte(segments[2]),
+			}
+		}
+	}
+	return nil
+}
+
+func (l *levelDBReceipts) getLookupKeysByIDs(ids []string, start, end string) []string {
+	result := []string{}
+	for _, id := range ids {
+		val, err := l.store.Get(id)
+		if err != nil {
+			log.Warnf("Failed to locate entry for key ID %s", id)
+			continue
+		} else {
+			result = append(result, string(val))
+		}
+	}
+	sort.Sort(sort.StringSlice(result))
+	// since our query searches in descending order, while sort.SearchString requires ascending order
+	// the "start" is the end, and "end" is the start
+	result = applyBound(result, end, start)
+	return result
+}
+
+func (l *levelDBReceipts) getLookupKeysByFromAndTo(from, to string, start, end string, limit int) []string {
+	if from != "" || to != "" {
+		fromKeys := []string{}
+		itr := l.store.NewIterator()
+		defer itr.Release()
+
+		if from != "" {
+			count := 0
```

Comment:
Seems like this block is a duplicate of the `to` block, with just the string different in the inserts, and `[]string` being a different set.
Could be split into a function to avoid the code duplication.